xapian-glib/xapian-term-iterator.cc

```diff
@@ -0,0 +1,321 @@
+ * @Title: XapianTermIterator
+ * @short_description: Iterate over terms
+ *
+ * #XapianTermIterator is a class that allows iterating over terms.
+ *
+ * A typical example is:
+ *
+ * |[<!-- language="C" -->
+ *   XapianTermIterator *iter = xapian_database_allterms (db);
+ *
+ *   while (xapian_term_iterator_next (iter))
+ *     g_print ("%s", xapian_term_iterator_get_term (iter));
+ * ]|
+ */
+
+class IteratorData {
+  public:
+    IteratorData (const IteratorData &aData) {
+      mCurrentInitialized = aData.mCurrentInitialized;
+      mCurrent = aData.mCurrent;
+    }
+
+    IteratorData (const Xapian::TermIterator &it) {
+      mCurrentInitialized = false;
+      mCurrent = it;
+    }
+
+    ~IteratorData () {
+    }
+
+    bool next () {
+      if (!mCurrentInitialized)
+        mCurrentInitialized = true;
+      else
+        ++mCurrent;
+
+      return Xapian::iterator_valid (mCurrent);
+    }
+
+    char *getTermName () {
+      if (!mCurrentInitialized)
+        return NULL;
+
+      const std::string &term = *mCurrent;
+
+      return g_strdup (term.c_str ());
+    }
+
+    unsigned int getWdf () {
+      if (!mCurrentInitialized)
+        return 0;
+
+      return mCurrent.get_wdf ();
+    }
+
+    unsigned int getTermFreq () {
+      if (!mCurrentInitialized)
+        return 0;
+
+      return mCurrent.get_termfreq ();
+    }
+
+    char * getDescription () {
+      const std::string &desc = mCurrent.get_description ();
+
+      return g_strdup (desc.c_str ());
+    }
+
+  private:
+
+    Xapian::TermIterator mCurrent;
+
+    bool mCurrentInitialized;
+};
+
+typedef struct {
+  IteratorData *data;
+} XapianTermIteratorPrivate;
+
+#define XAPIAN_TERM_ITERATOR_GET_PRIVATE(obj) \
+  ((XapianTermIteratorPrivate *) xapian_term_iterator_get_instance_private ((XapianTermIterator *) (obj)))
+
+G_DEFINE_TYPE_WITH_PRIVATE (XapianTermIterator, xapian_term_iterator, G_TYPE_OBJECT)
+
+static void
+xapian_term_iterator_finalize (GObject *gobject)
+{
+  XapianTermIteratorPrivate *priv = XAPIAN_TERM_ITERATOR_GET_PRIVATE (gobject);
+
+  delete priv->data;
+
+  G_OBJECT_CLASS (xapian_term_iterator_parent_class)->finalize (gobject);
+}
+
+static void
+xapian_term_iterator_class_init (XapianTermIteratorClass *klass)
+{
+  G_OBJECT_CLASS (klass)->finalize = xapian_term_iterator_finalize;
+}
+
+static void
+xapian_term_iterator_init (XapianTermIterator *self)
+{
+}
+
+/*< private >
+ * xapian_term_iterator_new:
+ * @it: a Xapian::TermIterator
+ *
+ * Creates a new #XapianTermIterator from the given `Xapian::TermIterator` instance.
+ *
+ * Returns: (transfer full): the newly created #XapianTermIterator
+ *
+ * Since: 1.4
+ */
+XapianTermIterator *
+xapian_term_iterator_new (const Xapian::TermIterator &it)
+{
+  XapianTermIterator *iter;
+  XapianTermIteratorPrivate *priv;
+
+  iter = static_cast<XapianTermIterator *> (g_object_new (XAPIAN_TYPE_TERM_ITERATOR, NULL));
+  priv = XAPIAN_TERM_ITERATOR_GET_PRIVATE (iter);
+  priv->data = new IteratorData (it);
```

Comment:
It seems clumsy that we have `IteratorData` at all, rather than just having this in `XapianTermIteratorPrivate`, but the latter seems to get allocated by glib somewhere behind a tangle of macros, so I guess it probably can't have C++ objects in it because the constructor likely won't get called.  Perhaps we could use C++ placement `new`.

Anyway, I'll change this.

BTW, I just copied this pattern from `xapian_mset_iterator_new` and `xapian_mset_iterator_init`, so I guess yet again I'll also fix the existing code to conform.